keras_nlp/src/models/phi3/phi3_rotary_embedding.py

```diff
@@ -0,0 +1,149 @@
+            rope frequencies when the `rope_scaling_type` is `"su"`. List must
+            be of length `hidden_dim//num_query_heads//2`. It is used when
+            `sequence_length` is smaller than `original_max_sequence_length`.
+            Defaults to `None`.
+        inverese_freq_long_factor List[float]: List of factors used to adjust
+            rope frequencies when the `rope_scaling_type` is `"su"`. List must
+            be of length `hidden_dim//num_query_heads//2`. It is used when
+            `sequence_length` is larger than `original_max_sequence_length`.
+            Defaults to `None`.
+        max_wavelength: int. The maximum angular wavelength of the sine/cosine
+            curves.
+
+    Call arguments:
+        inputs: The tensor inputs to apply the embedding to. This can have
+            any shape, but must contain both a sequence and feature axis. The
+            rotary embedding will be applied to `inputs` and returned.
+        start_index: An integer or integer tensor. The starting position to
+            compute the rotary embedding from. This is useful during cached
+            decoding, where each position is predicted separately in a loop.
+
+    References:
+     - [Phi-3-mini-128k-instruct original implementation](https://huggingface.co/microsoft/Phi-3-mini-128k-instruct/blob/0693e0b867d29e7318280ddd3ff9d5e66698f488/modeling_phi3.py#L142)
+    """
+
+    def __init__(
+        self,
+        max_sequence_length=4096,
+        original_max_sequence_length=4096,
+        inverese_freq_short_factor=None,
+        inverese_freq_long_factor=None,
+        max_wavelength=10000,
+        **kwargs
+    ):
+        super().__init__(max_wavelength=max_wavelength, **kwargs)
+        self.max_sequence_length = max_sequence_length
+        self.original_max_sequence_length = original_max_sequence_length
+
+        scaling_factor = (
+            self.max_sequence_length / self.original_max_sequence_length
+        )
+        if scaling_factor <= 1.0:
+            self.embedding_scaling_factor = 1.0
+        else:
+            self.embedding_scaling_factor = math.sqrt(
+                1
+                + math.log(scaling_factor)
+                / math.log(self.original_max_sequence_length)
+            )
+
+        if inverese_freq_short_factor is not None:
+            self.inverese_freq_short_factor = ops.convert_to_tensor(
+                inverese_freq_short_factor,
+                dtype="float32",
+            )
+        else:
+            self.inverese_freq_short_factor = None
+
+        if inverese_freq_long_factor is not None:
+            self.inverese_freq_long_factor = ops.convert_to_tensor(
+                inverese_freq_long_factor,
+                dtype="float32",
+            )
+        else:
+            self.inverese_freq_long_factor = None
+
+    def _compute_cos_sin_embedding(self, inputs, start_index=0, positions=None):
```

Comment:
This looks almost exactly the same as what's on the upstream. Is this possible to do just by overriding `self._get_inverse_freq(rotary_dim)`?

If so, would save a lot of code here.

Reply:
I will need to override `call()` also because `cos_embeddings` and `sine_embeddings` are also multiplied by a factor.
https://github.com/keras-team/keras-nlp/blob/0dff9f1eeda8dc37559c7b7a99514c1a8d469c17/keras_nlp/src/models/phi3/phi3_rotary_embedding.py#L129-L136